src/napari_graph/_base_graph.py

```diff
@@ -0,0 +1,652 @@
+        if edges_buffer[buffer_idx + 1] == tgt_node:
+            # skipping found edge from linked list
+            if prev_buffer_idx == _EDGE_EMPTY_PTR:
+                node2edges[src_node] = next_edge_idx
+            else:
+                edges_buffer[prev_buffer_idx + ll_edge_pos] = next_edge_idx
+
+            # clean up not necessary but good practice
+            edges_buffer[buffer_idx : buffer_idx + edge_size] = _EDGE_EMPTY_PTR
+
+            edges_buffer[buffer_idx + ll_edge_pos] = empty_idx
+            return idx
+        # moving to next edge
+        idx = next_edge_idx
+        prev_buffer_idx = buffer_idx
+
+    raise ValueError("Found an invalid edge at edge removal")
+
+
+@njit(inline='always')
+def _iterate_edges(
+    edge_ptr_indices: np.ndarray,
+    edges_buffer: np.ndarray,
+    edge_size: int,
+    ll_edge_pos: int,
+) -> typed.List:
+    """Iterate over the edges linked lists given their starting edges.
+       It returns list of multiplicity 2, where each pair is an edge.
+
+    Parameters
+    ----------
+    edge_ptr_indices : np.ndarray
+        Array of starting indices.
+    edges_buffer : np.ndarray
+        Edges buffer.
+    edge_size : int
+        Size of the edges on the buffer. It should be inlined when compiled.
+    ll_edge_pos : int
+        Position (shift) of the edge linked list on the edge buffer. It should
+        be inlined when compiled.
+
+    Returns
+    -------
+    typed.List
+        List of lists edges, adjacent nodes are at indices (k, k+1) such that k
+        is even.
+    """
+    edges_list = typed.List()
+
+    for idx in edge_ptr_indices:
+        edges = typed.List.empty_list(types.int64)
+        edges_list.append(edges)
+
+        while idx != _EDGE_EMPTY_PTR:
+            buffer_idx = idx * edge_size
+            edges.append(edges_buffer[buffer_idx])  # src
+            edges.append(edges_buffer[buffer_idx + 1])  # tgt
+            idx = edges_buffer[buffer_idx + ll_edge_pos]
+
+    return edges_list
+
+
+@njit
+def _create_world2buffer_map(world_idx: np.ndarray) -> typed.Dict:
+    """Fills world indices to buffer indices mapping."""
+    world2buffer = typed.Dict.empty(types.int64, types.int64)
+
+    for i in range(world_idx.shape[0]):
+        world2buffer[world_idx[i]] = i
+
+    return world2buffer
+
+
+@njit(parallel=True)  # TODO: benchmark if parallel is worth it
+def _vmap_world2buffer(
+    world2buffer: typed.Dict, world_idx: np.ndarray
+) -> typed.Dict:
+    """Maps world indices to buffer indices."""
+    buffer_idx = np.empty(world_idx.shape[0], dtype=types.int64)
+    for i in prange(world_idx.shape[0]):
+        buffer_idx[i] = world2buffer[world_idx[i]]
+    return buffer_idx
+
+
+class BaseGraph:
+    """Abstract base graph class.
+
+    Parameters
+    ----------
+    n_nodes : int
+        Number of nodes to allocate to graph.
+    ndim : int
+        Number of spatial dimensions of graph.
+    n_edges : int
+        Number of edges of the graph.
```

Comment:
I agree with you, since we will only support spatial graphs we need to provide the coordinates, so it could be `Graph(my_edges_sparse_matrix, node_indexed_coordinates_dataframe)`. Using a pandas data frame with a sparse matrix adds some flexibility with the indexing, rather than having a coordinate numpy array.

Reply:
Hi @JoOkuma and @jni, was wondering where we stand on this topic? Currently, napari-graph stores `coords`. Would you rather move this to napari's Graph layer, @jni?